plugins/catalog-backend-module-bitbucket-cloud/config.d.ts

```diff
@@ -59,6 +59,14 @@ export interface Config {
              * (Optional) TaskScheduleDefinition for the discovery.
              */
             schedule?: SchedulerServiceTaskScheduleDefinitionConfig;
+            /**
+             * (Optional) At what level discovery should take place, affecting Bitbucket Cloud API limits.
+             *
+             * Possible values:
+             * - `workspace` (default): 1 API call per workspace, limited to 900 repositories per workspace.
+             * - `project`: 1 API call per project, limited to 900 repositories per project.
+             */
+            level?: 'workspace' | 'project';
```

Comment:
I wonder whether we want to add this to the interface and provide the choice.

While the `project` option will utilize more API calls, is it that high to require this?

Also, the explanation is not fully correct. It is not necessarily `1 API call per workspace/project`.

Each paginated API request has a page length of 100 that results in `ceil(numRepos/pagelen)` pages. With 900 repositories this would have been 9 API calls (workspace).

For up to 100 projects, you will require 1 additional API call to get these. In general, you have `ceil(numProjects/pagelen)` API calls. How many you have depends on how you use these (e.g., per team, per product area, per ...).
For the regular cases, it might be 1 API call extra.

Now, you will have more API calls to get all repositories than you had with the workspace option (`ceil(num_repos / pagelen)`) as you very likely will have a non-empty page per project vs one overall. You would need the sum of `ceil(num_repo_project_i / pagelen)`.

For 900 repos, 9 projects, and equal distribution (100 each), you would end up with 10 API calls in total.
If 5 of them have 2 pages, and the other 4 have non-full single pages, you end up with 15 API calls in total.

For 900 repos, 20 projects, and equal distribution (45 each), you would end up with 21 API calls.

Of course, all of these are higher than the original 9 API calls. However, considering the API rate limits, these differences might not be super relevant.

Considering the constraints and instabilities of the API used, I wonder whether we need to find a better setup.

Reply:
That's a good point @pjungermann. It should have been `Minimum 1 API call per workspace/project`. 😄

I'll strip out the configurability and we'll go from there.

Reply:
🤯